[#283] elaboration on @Status method when network error happens (#289)
eclipse/microprofile-lra: spec/src/main/asciidoc/microprofile-lra-spec.adoc

or `@Complete` method invocation then it MAY report `410 Gone`
HTTP status code or in the case of non-JAX-RS method returning
<<source-ParticipantStatus,ParticipantStatus>> `null`.

Let's provide some intuition on this matter with the following example.

> Let's have an LRA participant annotated with `@Compensate` and `@Status`.

Reviewer: ```suggestion
Let's have an LRA participant containing methods annotated with `@Compensate` and `@Status`.
```

The LRA is cancelled and the LRA implementation tries to invoke the participant's `@Compensate` method.
At that time a network failure happens. The LRA implementation receives a network error as the result of the call.

> Now it cannot be sure if the `@Compensate` call pass through and participant has already compensated the work

Reviewer: ```suggestion
Now it cannot be sure if the `@Compensate` call passed through and participant has already compensated the work
```

and only the response was lost. Or if the `@Compensate` call has not reached the participant at all.

> As the participant defines the `@Status` method the LRA implementation will invoke it to find the participant's true status

Reviewer: ```suggestion
As the participant defines the `@Status` method the LRA implementation MUST invoke it to find the participant's true status
```

(invocation can be processed several times until succeeds).
The participant's `@Status` method is expected to indicates whether the `@Compensate` was successfully processed.

> * If the participant processed the `@Compensate` call it may <<jaxrs-participant-methods,forget about the LRA identification>> already.

Reviewer: ```suggestion
* If the participant processed the `@Compensate` call it may <<jaxrs-participant-methods, has already forgotten about the LRA identification>>.
```

  Then for JAX-RS participant, the response is `410 Gone`. For non-JAX-RS participant, the method returns `null`.

>   (The implementation MUST assume that all required actions are performed and the compensate invocation **is not** repeated.)

Reviewer: I would also add MUST NOT

* If the participant compensated successfully but it has not erased the notion of the LRA identity yet it may return
  <<source-ParticipantStatus,ParticipantStatus>> of `Compensated`. For JAX-RS participant it's a body of the response of return code `200`

>   for the non-JAX-RS participant, it's the returned value of the method. The compensate invocation **is not** repeated.

Reviewer: In non-jax-rs cases, there are several ways of how to return Compensated. Maybe point to the relevant sections, please?

* If the participant compensated but failed during the compensation then `FailedToCompensate` status may be returned.
  The LRA implementation then has to <<the-model,log sufficient information>> in such situation.
  The compensate invocation will **not be** repeated.
* If the participant has not yet compensated (ie. the initial invocation of `@Compensate` did not go through)
  then the `ParticipantStatus` will indicate the participant status with `Active`.

>   The implementation **repeats** the invocation of the `@Compensate` method.

Reviewer: MUST repeat

or `@Complete` method invocation then it MAY report `410 Gone`
HTTP status code or in the case of non-JAX-RS method returning
<<source-ParticipantStatus,ParticipantStatus>> `null`.

Let's provide some intuition on this matter with the following example.

Let's have an LRA participant containing methods annotated with `@Compensate` and `@Status`.
The LRA is cancelled and the LRA implementation tries to invoke the participant's `@Compensate` method.
At that time a network failure happens. The LRA implementation receives a network error as the result of the call.
Now it cannot be sure if the `@Compensate` call passed through and participant has already compensated the work

> and only the response was lost. Or if the `@Compensate` call has not reached the participant at all.

Reviewer: ```suggestion
and only the response was lost or if the `@Compensate` call has not reached the participant at all.
```

As the participant defines the `@Status` method the LRA implementation MUST invoke it to find the participant's true status
(invocation can be processed several times until succeeds).

> The participant's `@Status` method is expected to indicates whether the `@Compensate` was successfully processed.

Reviewer: ```suggestion
The participant's `@Status` method is expected to indicate whether the `@Compensate` was successfully processed.
```

* If the participant processed the `@Compensate` call it may <<jaxrs-participant-methods, has already forgotten about the LRA identification>>.
  Then for JAX-RS participant, the response is `410 Gone`. For non-JAX-RS participant, the method returns `null`.

>   (The implementation MUST assume that all required actions are performed and the compensate invocation MUST NOT be repeated.)

Reviewer 2: In some network setups requests are automatically replayed (MicroProfile FT for example) so I would not be this definitive.

Reviewer: These invocations are performed by the implementation so I wouldn't expect technologies like FT would be in place. This MUST be definitive because we give a guarantee with the Status method that Compensate doesn't need to be idempotent and thus it shouldn't be replayed.

Reviewer 2: You cannot guarantee that they won't be replayed.

Author: I agree with the @mmusgrov  that we can't guarantee the replay does happen. The text should be written from the perspective of the specification. I rephrased it.

As we don't know if the system uses some replay processing framework and as the message order can't be ensured (like compensate http call is invoked, there is some failure in the network and the call is buffered on some router, the initial invocation may timed-out (or coordinator crashes) and the status+compensate invocation is repeated, then suddenly the initial http request could be processed on the router and the duplicity occurs).

I didn't want to make the text complex initially, now I added a note from this perspective at the end.

Author: After discussing and re-thinking the point I'm leaving just the initial text without any additional note.

My perspective remains here - with status method the implementation has to guarantee that the invocation is not  repeated. The text is already rewritten in this regard.
I agree with @xstefank that specification needs to be definitive in this regard.

I'm in agreement with @mmusgrov's comment (if I have the correct understanding) that specification can't guarantee the replay does not happen. But the specification must guarantee that it's not replayed by the implementation. As mentioned, the text was rewritten in this regard.

* If the participant compensated successfully but it has not erased the notion of the LRA identity yet it may return

>   <<source-ParticipantStatus,ParticipantStatus>> of `Compensated`. For JAX-RS participant it's a body of the response of return code `200`

Reviewer 2: Remark: HTTP 410 GONE means the same thing.

Author: Yeap, thanks for the remark.

>   for the non-JAX-RS participant, one option is returning the value from the method. For other options to return `Compensated` consult the section <<non-jaxrs-participant-methods>>.

Reviewer 2: This sentence is difficult for an English reader to parse. If other people are able to parse it okay then please ignore this comment. But I would say that too much explanatory and or duplicated text can give the reader the impression that he is missing some subtle meaning.

Reviewer: agreed,  `for the non-JAX-RS participant case, consult the section <<non-jaxrs-participant-methods>>` may be enough?

Author: I changed the text.

>   The invocation of `@Compensate` method *is not* repeated.

Reviewer 2: Is the *is not* meant to conform to RFC 2119?

Reviewer: +1 MUST NOT be repeated

Reviewer 2: SHOULD NOT might be more implementable

Author: ok

> * If the participant compensated but failed during the compensation then `FailedToCompensate` status may be returned.

Reviewer 2: Is this not adequately explained elsewhere. If you include this here then you should also include all of the other cases (which I don't think you do) which have already been described in the spec.

Author: Agree. I elaborated the text.

>   The LRA implementation then has to <<the-model,log sufficient information>> in such situation.

Reviewer 2: Same comment applies here too.

Author: What is the missing piece here? (The details on the behaviour can be checked in the referred section.)

>   The compensate invocation will *not be* repeated.

Reviewer 2: You have just stated this behaviour on line 1169 and also on line 1173.

Author: For the clear information it's good to provide precise declaration of the behaviour. I think the text should state this here again.

* If the participant has not yet compensated (ie. the initial invocation of `@Compensate` did not go through)

>   then the `ParticipantStatus` will indicate the participant status with `Active`.

Reviewer 2: Aren't we just repeating what we already have in the state model.

Reviewer: This one should stay. I agree that the 'FailedToCompensate' section can be removed. But for the example purpose, this was exactly what I was looking for when creating the issue.

Author: I think the `FailedToCompensate` should NOT be removed. The point is to provide whole spectrum of the behaviours. It's essential idea of this example.

>   The implementation MUST repeat the invocation of the `@Compensate` method.

Reviewer 2: I agree but it does not read very well. Also this information should already be present elsewhere in the specification.

Author: You are more than welcome to provide me a suggestion for fixing my stylistic incompetence :-)

As the participant defines the `@Status` method the LRA implementation MUST invoke it to find the participant's true status
(invocation can be processed several times until succeeds).
The participant's `@Status` method is expected to indicate whether the `@Compensate` was successfully processed.

* If the participant processed the `@Compensate` call it may <<jaxrs-participant-methods, has already forgotten about the LRA identification>>.
  Then for JAX-RS participant, the response is `410 Gone`. For non-JAX-RS participant, the method returns `null`.
  The implementation MUST assume that all required actions are performed and the implementation MUST NOT repeat the invocation.
* If the participant compensated successfully but it has not erased the notion of the LRA identity yet it may return
  <<source-ParticipantStatus,ParticipantStatus>> of `Compensated`. For JAX-RS participant it's a body of the response of return code `200`.
  For the non-JAX-RS participant case, consult the section <<non-jaxrs-participant-methods>> (one option is to provide `Compensated` as the method return value).

>   The implementation MUST NOT repeat the invocation.

Reviewer: here is not clear whic invocation you mean - status or compensate. Status can be replayed but it doesn't make sense.

Author: I updated the text to clarified that's the `@Compensate` method.
